File: Assignment 2/dilsin_amirtharajah_assignment2.py
import re  # regular expressions


def number_check(s): # s is text
    pattern = r'^[0-9]+(\.[0-9]+)?$' # ^ start, digits, optional .digits, $ end
    return re.match(pattern, s) is not None # returns match object/none

while True: # loop
    first_input = input("Enter first number: ")  
    if number_check(first_input):                
        break                                    
    else:
        print("Error: Input must be a valid number.")  

while True:
    second_input = input("Enter second number: ")
    if number_check(second_input):
        break
    else:
        print("Error: Input must be a valid number.")

num1 = float(first_input) # convert text to decimal
num2 = float(second_input)


add_count = 0
sub_count = 0
mul_count = 0
div_count = 0
total_operations = 0
largest = 0.0


while True:

    print("\nChoose an option:\n"
          "1) Add two numbers\n"
          "2) Subtract two numbers\n"
          "3) Multiply two numbers\n"
          "4) Divide two numbers\n"
          "5) Show summary\n"
          "6) Reset summary\n"
          "7) Quit")
    choice_str = input("Choice: ")           

    if not re.match(r"^[1-7]$", choice_str): # ^ start, [1-7], $ end
        print("Error: Choice must be an integer from 1 to 7.")
        continue                              

    choice = int(choice_str) # text to number

    if choice == 1:                      
        result = num1 + num2
        print("Result:", f"{result:.2f}") # :.2f for format with 2 decimals
        add_count += 1 # increment 1
        total_operations += 1
        if result > largest:
            largest = result
    elif choice == 2:                  
        result = num1 - num2
        print("Result:", f"{result:.2f}")
        sub_count += 1
        total_operations += 1
        if result > largest:
            largest = result
    elif choice == 3:                  
        result = num1 * num2
        print("Result:", f"{result:.2f}")
        mul_count += 1
        total_operations += 1
        if result > largest:
            largest = result
    elif choice == 4:                  
        if num2 == 0:
            print("Error: Division by zero.")
        else:
            result = num1 / num2
            print("Result:", f"{result:.2f}")
            div_count += 1
            total_operations += 1
            if result > largest:
                largest = result
    elif choice == 5:                  
        print("--- Summary ---")
        print(f"Numbers entered: {num1:.2f} and {num2:.2f}")
        print(f"Total operations: {total_operations}")
        print(f"Largest result: {largest:.2f}")
        print(f"Additions: {add_count}")
        print(f"Subtraction: {sub_count}")
        print(f"Multiplications: {mul_count}")
        print(f"Divisions: {div_count}")
    elif choice == 6:                  
        add_count = 0
        sub_count = 0
        mul_count = 0
        div_count = 0
        total_operations = 0
        largest = 0.0
        print("Summary reset.")
    elif choice == 7:                  
        print("--- Summary ---")
        print(f"Numbers entered: {num1:.2f} and {num2:.2f}")
        print(f"Total operations: {total_operations}")
        print(f"Largest result: {largest:.2f}")
        print(f"Additions: {add_count}")
        print(f"Subtraction: {sub_count}")
        print(f"Multiplications: {mul_count}")
        print(f"Divisions: {div_count}")
        break   

File: Assignment 2/dilsin_amirtharajah_assignment2.cpp
#include <iostream>   
#include <string>
#include <iomanip> // decimals
#include <regex> // for number checker
using namespace std; 

bool numberCheck(const string &s) { // pass by reference
    regex realNumberCheck("^[0-9]+(\\.[0-9]+)?$"); // ^ to start, [0-9]+ for digits, (\.[0-9]+)? for other digits
    return regex_match(s, realNumberCheck); // returns true if pass
}

int main() {

    string firstInput;  // first number
    string secondInput; // second number

    while (true) {  // loop
        cout << "Enter first number: ";        
        cin >> firstInput; // read input
        if (numberCheck(firstInput)) { // checking number
            break;                              
        } else {                                
            cout << "Error: Input must be a valid number." << endl; 
        }
    }

    
    while (true) {                              
        cout << "Enter second number: ";
        cin >> secondInput;
        if (numberCheck(secondInput)) {         
            break;
        } else {
            cout << "Error: Input must be a valid number." << endl;
        }
    }

    
    double num1 = stod(firstInput);  // string to double
    double num2 = stod(secondInput); // string to double

    
    int addCount = 0; // times add
    int subCount = 0; // times subtract
    int mulCount = 0; // times multiply
    int divCount = 0; // times divide
    int totalOperations = 0; // total operation
    double largest = 0;  // largest result

    
    while (true) { // loop
        
        cout << "\nChoose an option:\n"
             << "1) Add two numbers\n"
             << "2) Subtract two numbers\n"
             << "3) Multiply two numbers\n"
             << "4) Divide two numbers\n"
             << "5) Show summary\n"
             << "6) Reset summary\n"
             << "7) Quit\n"
             << "Choice: ";

        string choiceStr; // storing choice as text
        cin >> choiceStr;   

        regex menuCheck("^[1-7]$"); // regex for characters 1-7
        if (!regex_match(choiceStr, menuCheck)) { // regex false
            cout << "Error: Choice must be an integer from 1 to 7." << endl;
            continue;
        }

        int choice = stoi(choiceStr); // string to int
        cout << fixed << setprecision(2); // always show decimals and setprecision for 2 digits 

        
        if (choice == 1) {                      
            double result = num1 + num2; // add
            cout << "Result: " << result << endl; 
            addCount++;// increment by 1
            totalOperations++;
            if (result > largest) largest = result; // update result if biggest
        }
        else if (choice == 2) {                  
            double result = num1 - num2;
            cout << "Result: " << result << endl;
            subCount++; // increment by 1
             totalOperations++;
            if (result > largest) largest = result;
        }
        else if (choice == 3) {                  
            double result = num1 * num2;
            cout << "Result: " << result << endl;
            mulCount++; // increment by 1
            totalOperations++;
            if (result > largest) largest = result;
        }
        else if (choice == 4) {                  
            if (num2 == 0) {
                cout << "Error: Division by zero." << endl;
            } else {
                double result = num1 / num2;
                cout << "Result: " << result << endl;
                divCount++; // increment by 1
                totalOperations++;
                if (result > largest) largest = result;
            }
        }
        else if (choice == 5) {                  
            cout << "--- Summary ---" << endl;
            cout << "Numbers entered: " << num1 << " and " << num2 << endl;
            cout << "Total operations: " << totalOperations << endl;
            cout << "Largest result: " << largest << endl;
            cout << "Additions: " << addCount << endl;
            cout << "Subtraction: " << subCount << endl;
            cout << "Multiplications: " << mulCount << endl;
            cout << "Divisions: " << divCount << endl;
        }
        else if (choice == 6) {                  
            addCount = 0; // reset all to 0
            subCount = 0;
            mulCount = 0;
            divCount = 0;
            totalOperations = 0;
            largest = 0;
            cout << "Summary reset." << endl;
        }
        else if (choice == 7) {                  
            cout << "--- Summary ---" << endl;
            cout << "Numbers entered: " << num1 << " and " << num2 << endl;
            cout << "Total operations: " << totalOperations << endl;
            cout << "Largest result: " << largest << endl;
            cout << "Additions: " << addCount << endl;
            cout << "Subtraction: " << subCount << endl;
            cout << "Multiplications: " << mulCount << endl;
            cout << "Divisions: " << divCount << endl;
            break;      
        }
    }

    return 0; 
}
